### PR TITLE
Add Options Page

### PR DESCRIPTION
**Pull request by RichartE**
Made it possible to customize the browser extension with a few options:
- Highlight (on or off)
- Invert highlight (yes or no)
- Choose a highlight color

Also, added link to options page from popup.

### CODE REVIEW

**Comment by LuisaE** [comment]
Those are good ideas, but we would want to discuss it:
1. People are not clicking the popup, so no one will see a page linked from there.
2. Changing the color is great! However, inverting the highlight is confusing because 1. we don't change the text in the box; 2. What does it mean? It's unclear.
3. I feel strongly about this one: Turning off the highlight shouldn't be an available feature. We are Addition viewer so highlighting additions is a main feature, and generally, you turn off those - people are using it for a reason, and if they didn't want it they wouldn't install it. Keeping the highlight button that doesn't work when it is off is weird. 

TLTR: I think we should keep the color, but not on an extra page, in the popup.

@sukritsangvong @shirin601 @ducndh @etiennerichart

**Comment by shirin601** [comment]
I am not really sure the inversion would do? If you could elaborate, that would be great.

I agree with Luisa. Turning the highlight on and off should not be a feature -- people would not download it if they did not want to use it. Furthermore, through both iterations of the user study, no one has complained about the highlights being a hindrance while reading. So I do not think this feature would help/add anything.

Lastly, the color is interesting. One person in my user study told me that we should keep the highlight button and highlights in the same color so that users can subconsciously make the connection. That might be something to think about.

**Comment by sukritsangvong** [comment]
^same comments as above but i don't really think having the option to turn off highlight is that big of a deal though. I would vote against implementing it (as if you don't want the highlight, u can just disable the extension), but if it's already implemented, I wouldn't be against keeping it.

**Comment by sukritsangvong** [comment]
Tested it out. Looks pretty good. 
- I would prefer that the options be on the popup extension rather than on a separate page. 
- Not 100% sure what the purpose of the `Invert Highlight` is. 
- When disabling the highlights, I would expect to have something on the page saying that the highlight has been disabled, else it is a bit confusing to see that there is X number of changes but nothing was highlighted. Because of this, we should either make it clear that the highlight is turned off on the page or remove it.

**Comment by sukritsangvong** [comment]
I think the highlight is broken. Nothing changes when I click the highlight button.

**Comment by RichartE** [comment]
I see the highlights changing on my computer, is anyone else getting the same issue?

**Comment by sukritsangvong** [comment]
Maybe it has something to do with what @LuisaE mentioned in discord. I just saw that when I am on your branch, I am for some reason several commits behind and there is a huge merge conflict when I tried to pull (I didn't change anything from your branch). Might be something to do with force push (I normally just merge main into my branch when I want to update my branch. Force push tends to act weird, so I never use them)

Will try to resolve this and see if I still get the problem.

**Comment by sukritsangvong** [comment]
Okay, now I see all the changes you made. I think it was something trippy with the force push. I manually checked out to the latest commit you made. We should merge the color option into the main, but we should remove the limitation from this pr and put it into a new pr. We should merge the limitation thingy when we actually do have the limitation page. Having the "go to limitation page" in the main that doesn't get us anywhere seems weird.

I will approve after the limitation thingy is moved to a new pr.

File: wikichange/assets/content.css [hunk 1]
```diff
@@ -170,4 +172,21 @@ input[type="text"] {
...
+    border: 5px solid white;
+    border-top: 5px solid #3498db;
+    border-radius: 100%;
+    width: 15px;
+    height: 15px;
+    position: absolute;
+    left: 102%;
+    display: inline-block;
+    animation: spin 2s linear infinite;
+  }
```
Comment (LuisaE) [inline wikichange/assets/content.css:187]: Cool, thank you for cleaning it up!

File: wikichange/src/content/content.js [hunk 2]
```diff
@@ -313,12 +293,14 @@ getPageCreationDate(title).then((date) => {
...
-                oldRevisionDate
-            )} and the present day. Some of the changes were purely formatting or deletions and, therefore, are not highlighted.`;
-            document.getElementById("revisionDate").innerHTML = new_text;
-        });
+            const new_text = `We highlighted <span style="color: #468946; font-weight: 700;">${found_count}</span> changes which represent additions to the page between ${
+                getRevisionToClosestDateText(
+                    getRevisionPageLink(title, curRevisionId, oldRevisionId).replace(/\s/g, "_"),
+                    oldRevisionDate
+                )
+            } and the present day. Some of the changes were purely formatting or deletions and, therefore, are not highlighted. Visit our <a href="">limitations</a> page to learn more.`;
```
Comment (LuisaE) [inline wikichange/src/content/content.js]: Let's remove the limitations for now

File: wikichange/src/content/markPageChar.js [hunk 3]
```diff
@@ -12,17 +12,22 @@ const addChars = (start, end) => {
     for(let i = start; i <= end; i++) {
         toMark += `mark#mark-${i}, `;
     }
+    if(toMark.length > 3000){
```
Comment (LuisaE) [inline wikichange/src/content/markPageChar.js]: Can you add space between if ( and for (?

File: wikichange/src/content/markPageWord.js [hunk 4]
```diff
@@ -24,25 +24,32 @@ const addWords = (start, end) => {
             break;
         }
     }
-    toMark += intermediary.reduce((accumulator, index) => `${accumulator}, mark#mark-${index}`, `mark#mark-${prevStart}`);
+    toMark += intermediary.reduce((accumulator, index) => `${accumulator}mark#mark-${index}, `, `mark#mark-${prevStart}, `);
+    if(toMark.length > 3000){
```
Comment (LuisaE) [inline wikichange/src/content/markPageWord.js]: Same here

File: wikichange/src/options/options.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+//https://developer.chrome.com/docs/extensions/mv3/options/
+
+// Saves options to chrome.storage
```
Comment (LuisaE) [inline wikichange/src/options/options.js]: Can you change to the /** comments? So we keep everything the same

File: wikichange/src/options/options.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        // Update status to let user know options were saved.
+        const status = document.getElementById('status');
+        status.textContent = 'Options saved.';
+        setTimeout(function () {
+            status.textContent = '';
+        }, 750);
+    });
+}
+
+// Restores select box and checkbox state using the preferences
```
Comment (LuisaE) [inline wikichange/src/options/options.js]: Same here /**

File: wikichange/src/popup/popup.html [hunk 7]
```diff
@@ -5,22 +5,22 @@
...
           <p>Welcome to our comps project: An extension to add insights to your Wikipedia experience.</p>
           <a href="https://sukritsangvong.github.io/wp-browser-extension/" id="comps-website" class="btn btn-secondary btn-sm">Learn more!</a>
           <hr>
         </div>
-        <p>
           <ul>
             <li>Year: 2022-2023</li>
           </ul>
-        </p>
+          <a role="button" href="" id="options">Extension Options</a>
```
Comment (LuisaE) [inline wikichange/src/popup/popup.html]: Maybe center year too? It's a little bit weird as it is to have one at the center and another one being a bullet

**Review by LuisaE — CHANGES_REQUESTED** [review]
Small formating changes, other than that good to go, thank you!

File: wikichange/src/content/content.js [hunk 8]
```diff
@@ -313,12 +293,14 @@ getPageCreationDate(title).then((date) => {
...
-                oldRevisionDate
-            )} and the present day. Some of the changes were purely formatting or deletions and, therefore, are not highlighted.`;
-            document.getElementById("revisionDate").innerHTML = new_text;
-        });
+            const new_text = `We highlighted <span style="color: #468946; font-weight: 700;">${found_count}</span> changes which represent additions to the page between ${
+                getRevisionToClosestDateText(
+                    getRevisionPageLink(title, curRevisionId, oldRevisionId).replace(/\s/g, "_"),
+                    oldRevisionDate
+                )
+            } and the present day.`;
```
Comment (LuisaE) [inline wikichange/src/content/content.js]: We want to keep "Some of the changes were purely formatting or deletions and, therefore, are not highlighted.". The rest looks good, thanks

**Review by sukritsangvong — APPROVED** [review]
lgtm @LuisaE

**Review by LuisaE — APPROVED** [review]

**Pull request merged**